pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -54,6 +54,9 @@ const (
 	gceAffinityTypeClientIP = "CLIENT_IP"
 	// AffinityTypeClientIPProto - affinity based on Client IP and port.
 	gceAffinityTypeClientIPProto = "CLIENT_IP_PROTO"
+
+	operationPollInterval        = 3 * time.Second
+	operationPollTimeoutDuration = 7 * time.Minute
```

Comment:
> It's generally a violation of the GCE api contract for an operation to never go to DONE, but there are known edge cases where it can take 15-20 minutes or even more to get there.

I'm a big believer in defensive programming. Having a loop that can block indefinitely if the server returns the wrong thing or if there is a bug in the deserializer seems like a bad idea. I'm completely open to upping the limit to something very high... 30 minutes? Would you be ok with that?

> Does the pd retry code handle the case where there's an already-pending operation to detach a disk and GCE returns an error creating a new operation to do the same thing? If not, then retries may not be improving reliability.

The pd retry code would treat an already-pending operation as any other failure and would retry after some time. You're right that retries wouldn't improve reliability in that case. But they will eventually put the pod in a bad state, and generate events indicating that the pod is blocked because it is unable to attach the disk, because the attach timedout. Meaning the user eventually gets indication of what is going on, vs if this code blocked indefinitely the pod would be stuck forever in pending state, with no indication to the user what is happening.